lib/hal/led.c

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright and related rights waived via CC0
+https://creativecommons.org/publicdomain/zero/1.0/
+*/
+
+#include <hal/led.h>
+#include <xboxkrnl/xboxkrnl.h>
+#include <assert.h>
+
+#define LED_MODE_AUTO   0x00
+#define LED_MODE_MANUAL 0x01
+
+#define SMC_REG_LEDMODE 0x07
+#define SMC_REG_LEDSEQ  0x08
+
+unsigned int XGetLEDMode(void) {
+  ULONG ret = 0;
+  HalReadSMBusValue(0x20, SMC_REG_LEDMODE, FALSE, &ret);
+  return ret;
+}
+
+unsigned int XGetLEDColorScheme(void) {
+  ULONG ret = 0;
+  if (XGetLEDMode() == LED_MODE_MANUAL) {
+    HalReadSMBusValue(0x20, SMC_REG_LEDSEQ, FALSE, &ret);
+  }
+  return ret;
+}
```

Comment:
This makes the API asymmetric.

I'd rather have it be `XGetManualLED(unsigned int *t1, [...])` and a return value wether it was successful.

For the auto/manual mode it can be a `bool XIsAutoLED()` (and maybe a negating `bool XIsManualLED() { return !XIsAutoLED(); }`).

Maybe someone else has design thoughts?

*I also don't think this "getter" is really necessary, as the application is normally under full control of LED (so it knows the state), and it can always fall back to lower-level APIs if it needs niche features.*

Reply:
Other idea for API: `void XResetLED()` / `void XSetCustomLED(unsigned int t1, [...])` and `bool XGetCustomLED(unsigned int* t1, [...])` which returns an error to indicate LED wasn't even in custom mode.

So instead of naming 2 modes, just have 1 default mode and 1 named mode (custom).

Reply:
Agree. Don't see much of a use case for this getter.